File: ruby-app/controllers/api_keys.rb
module Controllers
  class ApiKeys
    def initialize(reloop_client)
      @client = reloop_client
    end

    # 1. POST /api/api-keys - Create API Key
    def create(body)
      name = body["name"]
      raise ArgumentError, "Name is required" if name.nil? || name.strip.empty?

      result = @client.api_keys.create({ name: name })
      { status: 201, body: { success: true, data: result } }
    end

    # 2. GET /api/api-keys - List API Keys
    def list(params)
      limit = (params["limit"] || 10).to_i
      page = (params["page"] || 1).to_i

      result = @client.api_keys.list({ limit: limit, page: page })
      { status: 200, body: { success: true, data: result } }
    end

    # 3. GET /api/api-keys/:id - Get Key Details
    def get(id)
      result = @client.api_keys.get(id)
      { status: 200, body: { success: true, data: result } }
    end

    # 4. PATCH /api/api-keys/:id - Update / Rename Key
    def update(id, body)
      name = body["name"]
      raise ArgumentError, "Name is required" if name.nil? || name.strip.empty?

      result = @client.api_keys.update(id, { name: name })
      { status: 200, body: { success: true, data: result } }
    end

    # 5. POST /api/api-keys/:id/disable - Disable Key
    def disable(id)
      result = @client.api_keys.disable(id)
      { status: 200, body: { success: true, data: result } }
    end

    # 6. POST /api/api-keys/:id/enable - Enable Key
    def enable(id)
      result = @client.api_keys.enable(id)
      { status: 200, body: { success: true, data: result } }
    end

    # 7. POST /api/api-keys/:id/rotate - Rotate Key Secret
    def rotate(id)
      result = @client.api_keys.rotate(id)
      { status: 200, body: { success: true, data: result } }
    end

    # 8. DELETE /api/api-keys/:id - Delete Key
    def delete(id)
      result = @client.api_keys.delete(id)
      { status: 200, body: { success: true, data: result } }
    end
  end
end

File: rust-app/src/controllers/api_keys.rs
use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use reloop_email::{models::*, ReloopClient};
use serde::Deserialize;
use serde_json::json;
use std::sync::Arc;

#[derive(Deserialize)]
pub struct CreateKeyPayload {
    pub name: String,
}

#[derive(Deserialize)]
pub struct UpdateKeyPayload {
    pub name: String,
}

#[derive(Deserialize)]
pub struct PaginationQuery {
    pub page: Option<i32>,
    pub limit: Option<i32>,
}

// 1. POST /api/api-keys - Create API Key
pub async fn create_key(
    State(client): State<Arc<ReloopClient>>,
    Json(payload): Json<CreateKeyPayload>,
) -> impl IntoResponse {
    let params = CreateApiKeyParams {
        name: payload.name,
        enabled: Some(true),
        rate_limit_enabled: Some(true),
    };

    match client.api_keys().create(params).await {
        Ok(result) => (
            StatusCode::CREATED,
            Json(json!({ "success": true, "data": result })),
        ),
        Err(err) => (
            StatusCode::BAD_REQUEST,
            Json(json!({ "success": false, "error": err.to_string() })),
        ),
    }
}

// 2. GET /api/api-keys - List API Keys
pub async fn list_keys(
    State(client): State<Arc<ReloopClient>>,
    Query(query): Query<PaginationQuery>,
) -> impl IntoResponse {
    let params = ApiKeyListParams {
        page: query.page.or(Some(1)),
        limit: query.limit.or(Some(10)),
        ..Default::default()
    };

    match client.api_keys().list(Some(params)).await {
        Ok(result) => (
            StatusCode::OK,
            Json(json!({ "success": true, "data": result })),
        ),
        Err(err) => (
            StatusCode::BAD_REQUEST,
            Json(json!({ "success": false, "error": err.to_string() })),
        ),
    }
}

// 3. GET /api/api-keys/:id - Get Key Details
pub async fn get_key(
    State(client): State<Arc<ReloopClient>>,
    Path(id): Path<String>,
) -> impl IntoResponse {
    match client.api_keys().get(&id).await {
        Ok(result) => (
            StatusCode::OK,
            Json(json!({ "success": true, "data": result })),
        ),
        Err(err) => (
            StatusCode::BAD_REQUEST,
            Json(json!({ "success": false, "error": err.to_string() })),
        ),
    }
}

// 4. PATCH /api/api-keys/:id - Update / Rename Key
pub async fn update_key(
    State(client): State<Arc<ReloopClient>>,
    Path(id): Path<String>,
    Json(payload): Json<UpdateKeyPayload>,
) -> impl IntoResponse {
    let params = UpdateApiKeyParams {
        name: Some(payload.name),
        enabled: Some(true),
    };

    match client.api_keys().update(&id, params).await {
        Ok(result) => (
            StatusCode::OK,
            Json(json!({ "success": true, "data": result })),
        ),
        Err(err) => (
            StatusCode::BAD_REQUEST,
            Json(json!({ "success": false, "error": err.to_string() })),
        ),
    }
}

// 5. POST /api/api-keys/:id/disable - Disable Key
pub async fn disable_key(
    State(client): State<Arc<ReloopClient>>,
    Path(id): Path<String>,
) -> impl IntoResponse {
    match client.api_keys().disable(&id).await {
        Ok(result) => (
            StatusCode::OK,
            Json(json!({ "success": true, "data": result })),
        ),
        Err(err) => (
            StatusCode::BAD_REQUEST,
            Json(json!({ "success": false, "error": err.to_string() })),
        ),
    }
}

// 6. POST /api/api-keys/:id/enable - Enable Key
pub async fn enable_key(
    State(client): State<Arc<ReloopClient>>,
    Path(id): Path<String>,
) -> impl IntoResponse {
    match client.api_keys().enable(&id).await {
        Ok(result) => (
            StatusCode::OK,
            Json(json!({ "success": true, "data": result })),
        ),
        Err(err) => (
            StatusCode::BAD_REQUEST,
            Json(json!({ "success": false, "error": err.to_string() })),
        ),
    }
}

// 7. POST /api/api-keys/:id/rotate - Rotate Key Secret
pub async fn rotate_key(
    State(client): State<Arc<ReloopClient>>,
    Path(id): Path<String>,
) -> impl IntoResponse {
    match client.api_keys().rotate(&id).await {
        Ok(result) => (
            StatusCode::OK,
            Json(json!({ "success": true, "data": result })),
        ),
        Err(err) => (
            StatusCode::BAD_REQUEST,
            Json(json!({ "success": false, "error": err.to_string() })),
        ),
    }
}

// 8. DELETE /api/api-keys/:id - Delete Key
pub async fn delete_key(
    State(client): State<Arc<ReloopClient>>,
    Path(id): Path<String>,
) -> impl IntoResponse {
    match client.api_keys().delete(&id).await {
        Ok(result) => (
            StatusCode::OK,
            Json(json!({ "success": true, "data": result })),
        ),
        Err(err) => (
            StatusCode::BAD_REQUEST,
            Json(json!({ "success": false, "error": err.to_string() })),
        ),
    }
}
